baseplate/sidecars/live_data_watcher.py

```diff
@@ -32,6 +36,41 @@ def __init__(self, dest: str, owner: int, group: int, mode: int):
         self.group = group
         self.mode = mode
 
+    @staticmethod
+    def fetch_data_from_url(url: str) -> Optional[str]:
+        data = None
+        try:
+            # Fetch the data from the url as bytes.
+            data = requests.get(url).content
+        except requests.exceptions.RequestException as e:
+            logger.exception(e)
+            return None
+        return data
+
+    @staticmethod
+    def get_data_to_write(json_data: dict, data: bytes) -> Optional[bytes]:
+        # Check if we have a JSON in this special format:
+        # data = {
+        #    "live_data_watcher_load_type": str
+        #    "data": str
+        #    "md5_hashed_data": str
+        # }
+        # If the load type is 'http', this format is an indication that we support
+        # downloading the contents of files uploaded to S3, GCS, etc when provided
+        # with an accessible URL.
+        if json_data.get("live_data_watcher_load_type") == "http":
+            # Only write the data if we actually managed to fetch its contents.
+            url = json_data.get("data")
+            if url is None:
+                logger.exception("No url found in zk source JSON node.")
+                return None
+            url_data = NodeWatcher.fetch_data_from_url(url)
+            if url_data is None:
+                return None
+            return url_data
+        else:
+            return data
+
```

Comment:
This code appears to imply that configuration files would be available over HTTP unauthenticated. This makes me pretty uncomfortable for two reasons:

1. We have no way of authenticating that the server is trusted
2. We have little or no auditability into who accessed configuration and what response they received

Generally, I suggest treating configuration like source code, so it should be equally as locked down.

Personally, I would suggest integrating with EC2 and GCP directly. This would allow us to bootstap with their auth methods. It also gives us access to their audit logs if we need to verify the integrity of the configurations. Both of these platforms also have ways that you can watch for changes  on objects, which is probably preferred to polling,

Reply:
I think the implication is that auth should be handled by service account so that you don't have to do anything special in the code, instead of "just make them publicly readable".

Reply:
I would expect that if auth was being delegated to a service account we would be using a client to connect that would deal with auth. For example, the S3 client would add auth details to a request.

Generally I would agree that application owners shouldn't need to be directly concerned with authentication to access live configurations.

Reply:
I would suggest that instead of `"live_data_watcher_load_type": "http"` we use `"s3"` instead and use boto to download the file.

Reply:
Actually I see what you both are saying.

@KTAtkinson maybe instead of `requests.get` we can actually just use the `boto3` s3 client directly. This would mean we probably want to specify the `live_data_watcher_load_type` as `S3` and NOT `url`. This way it's more secure and explicit that we're using S3 client.

Reply:
Updated with new secure flow.